File: trie/LC208.cpp
#include <bits/stdc++.h>

using namespace std;

struct Node {
    Node* next[26];
    bool term;
};


class Trie {
public:
    Node* root;
    Trie() {
        root = new Node();
    }
    
    void insert(string word) {
        Node* cur = root;
        for(char ch: word){
            ch -= 'a';
            if (cur->next[ch] == nullptr) {
                cur->next[ch] = new Node();
            }
            cur = cur->next[ch];
        }
        cur->term = true;
    }

    Node* traverse(string s) {
        Node* cur = root;
        for (char ch: s) {
            ch -= 'a';
            if (cur->next[ch] == nullptr) {
                return nullptr;
            }
            cur = cur->next[ch];
        }
        return cur;
    }
    
    bool search(string word) {
        Node* node = traverse(word);
        return node != nullptr && node->term;
    }
    
    bool startsWith(string prefix) {
        Node* node = traverse(prefix);
        return node != nullptr;
    }
};

File: trie/LC208.py
# https://leetcode.cn/problems/implement-trie-prefix-tree/
from collections import defaultdict


class Node:
    __slots__ = ['next', 'term']

    def __init__(self) -> None:
        self.next = defaultdict(Node)
        self.term = False


class Trie:

    def __init__(self):
        self.root = Node()

    def insert(self, word: str) -> None:
        cur = self.root
        for ch in word:
            cur = cur.next[ch]
        cur.term = True

    def search(self, word: str) -> bool:
        cur = self.root
        for ch in word:
            if ch not in cur.next:
                return False
            cur = cur.next[ch]
        return cur.term

    def startsWith(self, prefix: str) -> bool:
        cur = self.root
        for ch in prefix:
            if ch not in cur.next:
                return False
            cur = cur.next[ch]
        return True
